### PR TITLE
Add `ReportBody` wrapper around `sgx_report_body_t`

### PR DESCRIPTION
**Pull request by nick-mobilecoin**

### CODE REVIEW

**Comment by nick-mobilecoin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #91** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/91" target="_blank">Graphite</a> 
    * **PR #86** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/86" target="_blank">Graphite</a> 
      * **PR #92** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/92" target="_blank">Graphite</a> 
        * **PR #103** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/103" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/103?utm_source=stack-comment).

File: core/types/src/report.rs [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+    MiscellaneousSelect,
+};
+use mc_sgx_core_sys_types::{
+    sgx_config_id_t, sgx_isvext_prod_id_t, sgx_isvfamily_id_t, sgx_measurement_t, sgx_prod_id_t,
+    sgx_report_body_t, sgx_report_data_t, SGX_CONFIGID_SIZE, SGX_HASH_SIZE,
+    SGX_ISVEXT_PROD_ID_SIZE, SGX_REPORT_BODY_RESERVED1_BYTES, SGX_REPORT_BODY_RESERVED2_BYTES,
+    SGX_REPORT_BODY_RESERVED3_BYTES, SGX_REPORT_BODY_RESERVED4_BYTES, SGX_REPORT_DATA_SIZE,
+};
+
+/// A measurement
```
Comment (nick-mobilecoin) [inline core/types/src/report.rs:15]: Up for suggestions on these transitive types.  Don't think it's ideal they're here but still flushing out where else they might be used.

**Review by jcape — CHANGES_REQUESTED** [review]
Recommend making `Measurement` an enum and using distinct types for MRENCLAVE and MRSIGNER hash types.

File: core/types/src/report.rs [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+    Measurement, sgx_measurement_t, m;
+}
+
+impl Default for Measurement {
+    fn default() -> Self {
+        Self(sgx_measurement_t {
+            m: [0; SGX_HASH_SIZE],
+        })
+    }
+}
```
[the remark this answers is not in the document]
Comment (nick-mobilecoin) [reply core/types/src/report.rs]: Brought in most of the Measurement logic from mc_attest_core

File: core/types/src/report.rs [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+new_type_accessors_impls! {
+    ConfigId, sgx_config_id_t;
+}
+
+impl Default for ConfigId {
+    fn default() -> Self {
+        Self([0; SGX_CONFIGID_SIZE])
+    }
+}
```
Comment (jcape) [inline core/types/src/report.rs:61]: Can this be derived?

File: core/types/src/report.rs [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    ReportBody, sgx_report_body_t;
+}
+
+// Implementing default to make it easier to pass an `sgx_report_body_t` to the
+// sgx functions.
+// ```
+// let mut report: sgx_report_body_t = Report::default().into()
+// let return_value = unsafe{ sgx_some_call(&report as *mut _) }
+// ```
+impl Default for ReportBody {
```
Comment (jcape) [inline core/types/src/report.rs:149]: Same Q: can this be derived?

File: core/types/src/macros.rs [hunk 5]
```diff
@@ -38,3 +40,139 @@ macro_rules! new_type_accessors_impls {
...
+            fn try_from(src: $crate::macros::Vec<u8>) -> core::result::Result<Self, Self::Error> {
+                Self::try_from(src.as_slice())
+            }
+        }
+
+        impl Default for $wrapper {
+            fn default() -> Self {
+                Self($inner { $fieldname: [0u8; $size] })
+            }
+        }
```
Comment (jcape) [inline core/types/src/macros.rs:98]: Can we remove this and just derive it?

**Review by jcape — APPROVED** [review]
LGTM, but ideally we are just `derive(Default)` where applicable---the underlying FFI types should all have default on them now anyways, right?

File: core/types/src/report.rs (hunk 4 above)
Comment (nick-mobilecoin) [reply core/types/src/report.rs:149]: Done as part of #115

File: core/types/src/macros.rs (hunk 5 above)
Comment (nick-mobilecoin) [reply core/types/src/macros.rs:98]: Done as part of #115

**Review by awygle — APPROVED** [review]

**Comment by nick-mobilecoin** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/103?utm_source=stack-comment).


Job Reference: C6ZyZ0Ggec8V0ksW3mYN

**Pull request merged**